### PR TITLE
Propagate room join errors to the UI

### PR DESCRIPTION
**Pull request by dbkr**
Dispatch so we can set the state in RoomViewStore. Show the error
when the room join fails (unsure if it's better to do this from
the component or the store). Remove unused joinError from roomview.

### CODE REVIEW

File: src/stores/RoomViewStore.js [hunk 1]
```diff
@@ -128,17 +137,41 @@ class RoomViewStore extends Store {
...
+            const ErrorDialog = sdk.getComponent("dialogs.ErrorDialog");
+            Modal.createDialog(ErrorDialog, {
+                title: _t("Failed to join room"),
+                description: msg,
+            });
+        });
+    }
+
+    _joinedRoom(payload) {
+        if (payload.room_id === this._state.roomId) {
```
Comment (lukebarnard1) [inline src/stores/RoomViewStore.js]: Evil! We must not update state conditionally but instead trust the payload!

File: src/stores/RoomViewStore.js [hunk 2]
```diff
@@ -128,17 +137,41 @@ class RoomViewStore extends Store {
...
             this._setState({
                 joining: false,
             });
-        }, (err) => {
+        }
+    }
+
+    _joinRoomError(payload) {
+        if (payload.room_id === this._state.roomId) {
```
Comment (lukebarnard1) [inline src/stores/RoomViewStore.js]: See above

**Comment by dbkr** [comment]
done!

File: src/stores/RoomViewStore.js [hunk 3]
```diff
@@ -128,16 +137,36 @@ class RoomViewStore extends Store {
...
             joining: true,
         });
-        MatrixClientPeg.get().joinRoom(this._state.roomId, payload.opts).then(
-        () => {
-            this._setState({
-                joining: false,
+        MatrixClientPeg.get().joinRoom(this._state.roomId, payload.opts).done(() => {
+            dis.dispatch({
+                action: 'joined_room',
+                room_id: this._state.roomId,
```
Comment (lukebarnard1) [inline src/stores/RoomViewStore.js]: This isn't used anymore

File: src/stores/RoomViewStore.js [hunk 4]
```diff
@@ -128,16 +137,36 @@ class RoomViewStore extends Store {
...
             });
         }, (err) => {
-            this._setState({
-                joining: false,
-                joinError: err,
+            dis.dispatch({
+                action: 'join_room_error',
+                room_id: this._state.roomId,
```
Comment (lukebarnard1) [inline src/stores/RoomViewStore.js]: See above

File: src/stores/RoomViewStore.js [hunk 5]
```diff
@@ -128,16 +137,36 @@ class RoomViewStore extends Store {
...
+                err: err,
             });
+            const msg = err.message ? err.message : JSON.stringify(err);
+            const ErrorDialog = sdk.getComponent("dialogs.ErrorDialog");
+            Modal.createDialog(ErrorDialog, {
```
Comment (lukebarnard1) [inline src/stores/RoomViewStore.js]: It feels quite wrong to create modals in the store. I would at least move this to `_onRoomViewStoreUpdate` of `RoomView`, but I am surprised that we weren't already doing this somewhere else...

File: src/stores/RoomViewStore.js (hunk 3 above)
Comment (dbkr) [inline src/stores/RoomViewStore.js]: This isn't, is it? Since we need to update `joining` to false.

File: src/stores/RoomViewStore.js (hunk 4 above)
Comment (dbkr) [inline src/stores/RoomViewStore.js]: Agreed that nothing reads this error, but if it's there then it should probably be kept up to date. We could remove the join error though (unless we need it in RoomView by moving the Modal there)

File: src/stores/RoomViewStore.js (hunk 5 above)
Comment (dbkr) [inline src/stores/RoomViewStore.js]: We were doing it in RoomView but it was removed - I wasn't sure whether the intention was to have the thing doing the joining display these errors or the view that caused the join: if the former, you'd get dialogs pop up even if you navigated away from the room view, if the latter, you could get multiple dialogs if multiple things were listening to the dispatch.

File: src/stores/RoomViewStore.js (hunk 3 above)
Comment (lukebarnard1) [inline src/stores/RoomViewStore.js]: But we don't need the room_id to do that

File: src/stores/RoomViewStore.js (hunk 4 above)
Comment (lukebarnard1) [inline src/stores/RoomViewStore.js]: I was referring to the `room_id`

File: src/stores/RoomViewStore.js (hunk 5 above)
Comment (lukebarnard1) [inline src/stores/RoomViewStore.js]: IRL we decided that the value added by moving the location of the dialog creation is probably quite low. 

The store should probably not be doing API calls and it should probably be dumb in that it should just accept state changes via dispatches that it trusts fully but that have originated elsewhere.

**Pull request merged**